### ISSUE TITLE
Examine why tor is opening many (20+) circuits; is this correct behavior?

### ISSUE THREAD

**Issue by mtigas** [OnionBrowser/iOS-OnionBrowser#226]
(example output from @tladesignz testing; i've added newlines before each `BUILT` output line)
```
24
BUILT $E135D5384EA42B4A771268CDA139D73D1C16A2C4~yutu,$1CB6DB698C49965D3C75A09B8D9EBFDD0E3B5603~kleinbus,$7BFB908A3AA5B491DA4CA72CCBEE0E1F2A939B55~sofia BUILD_FLAGS=NEED_CAPACITY PURPOSE=GENERAL TIME_CREATED=2019-12-10T12:34:01.39164748
BUILT $E135D5384EA42B4A771268CDA139D73D1C16A2C4~yutu,$F6E26FDF3A14117372856476339F878EEB8C2BB8~Nozel10,$DE465B3D36877D578627120DEF61C166D653FC4E~yuicat11 BUILD_FLAGS=IS_INTERNAL,NEED_CAPACITY PURPOSE=GENERAL TIME_CREATED=2019-12-10T13:26:44.03221327
BUILT $E135D5384EA42B4A771268CDA139D73D1C16A2C4~yutu,$EE4AF632058F0734C1426B1AD689F47445CA2056~angeltest7,$1297DB025DDDC46ADCB9D232E15B02E65272FE72~TPC2 BUILD_FLAGS=NEED_CAPACITY PURPOSE=GENERAL TIME_CREATED=2019-12-10T13:26:08.91619749
BUILT $E135D5384EA42B4A771268CDA139D73D1C16A2C4~yutu,$C4673051AF84961E55BAA1D25655CE5168BD70BE~Piratenpartei4,$EC80EECFCD599612C9A74678C8001146450141E8~fr0akatala BUILD_FLAGS=IS_INTERNAL,NEED_CAPACITY PURPOSE=GENERAL TIME_CREATED=2019-12-10T13:26:45.03660450
BUILT $E135D5384EA42B4A771268CDA139D73D1C16A2C4~yutu,$A4C83C2BEB4F1B31D40E80D5524C66A117D77C50~kiwirelay,$395D293263E9DEEC1874503D45B8543DABACC694~snukafly BUILD_FLAGS=IS_INTERNAL,NEED_CAPACITY PURPOSE=GENERAL TIME_CREATED=2019-12-10T13:26:46.03676530
BUILT $E135D5384EA42B4A771268CDA139D73D1C16A2C4~yutu,$511F66E40495E3B45EA21F0023DF714DB487257A~MLTorNode02,$0FB2EDAB99D10A2B708370E625323C76EB895FA2~niftyhazelmouse BUILD_FLAGS=NEED_CAPACITY PURPOSE=GENERAL TIME_CREATED=2019-12-10T13:26:22.68372531
BUILT $E135D5384EA42B4A771268CDA139D73D1C16A2C4~yutu,$A6E0F950CAC74AF867CA21976069529D6D4D3E30~HRMB,$4031460683AE9E0512D3620C2758D98758AC6C93~niftyeuropeanrabbit BUILD_FLAGS=NEED_CAPACITY PURPOSE=GENERAL TIME_CREATED=2019-12-10T13:26:22.95912532
BUILT $E135D5384EA42B4A771268CDA139D73D1C16A2C4~yutu,$0C3D5E19E3C75B505C8ACD26F89DCA2DF970553E~HORUS1,$DE847D94E78B2E560AB87D272DC90192D3144F17~kohlrabi BUILD_FLAGS=NEED_CAPACITY PURPOSE=GENERAL TIME_CREATED=2019-12-10T13:26:23.95957433
BUILT $E135D5384EA42B4A771268CDA139D73D1C16A2C4~yutu,$F65990C23AB017AC37226EB2AD3DA5A8EAE55D0C~binaery,$3A65D7E0FA8FC92AF99FB54CD8CA3276FBA33F03~Unnamed,$B5506D8E03F7BEAFA1A955DD3C39B38F830B2F4E~tor4kryptonit BUILD_FLAGS=IS_INTERNAL,NEED_CAPACITY PURPOSE=HS_CLIENT_HSDIR HS_STATE=HSCI_CONNECTING TIME_CREATED=2019-12-10T13:26:30.89451634
BUILT $E135D5384EA42B4A771268CDA139D73D1C16A2C4~yutu,$73F89EF2EE0F5C54064AC52BE9C111349CD9E8C4~bbtrotro,$3DC748C2DFF1F8E9DA5626BB3CBAAFF31D706C10~Fussy BUILD_FLAGS=IS_INTERNAL,NEED_CAPACITY PURPOSE=HS_CLIENT_REND HS_STATE=HSCR_JOINED REND_QUERY=3g2upl4pq6kufc4m TIME_CREATED=2019-12-10T13:26:30.98581735
BUILT $E135D5384EA42B4A771268CDA139D73D1C16A2C4~yutu,$066735D43FB476FA64578C5D829E176DF53E5A00~FreifunkIbbenbueren,$9E1E4F5B5F94812D02C4D18CB4086CE71CA5C614~torpidsDEhetzner1 BUILD_FLAGS=IS_INTERNAL,NEED_CAPACITY PURPOSE=HS_CLIENT_REND HS_STATE=HSCR_ESTABLISHED_IDLE REND_QUERY=3g2upl4pq6kufc4m TIME_CREATED=2019-12-10T13:26:31.98795137
BUILT $E135D5384EA42B4A771268CDA139D73D1C16A2C4~yutu,$25990FC54D7268C914170A118EE4EE75025451DA~TotorBE1,$A1B7C1F83E5DA4376FD2E1DBA7AC28A71AB332AD~arnall,$03910F285A33F365838EC66EF2C2EF754D046760~operator BUILD_FLAGS=IS_INTERNAL,NEED_CAPACITY PURPOSE=HS_CLIENT_INTRO HS_STATE=HSCI_CONNECTING REND_QUERY=3g2upl4pq6kufc4m TIME_CREATED=2019-12-10T13:26:32.98894139
BUILT $E135D5384EA42B4A771268CDA139D73D1C16A2C4~yutu,$3239007CE1FB2ECDFDF2067DF23B949295DC5EF6~TOR2DFN01a,$645DE9BF7A2E858F8A6B45F1F530371176D0238A~coco BUILD_FLAGS=IS_INTERNAL,NEED_CAPACITY PURPOSE=GENERAL TIME_CREATED=2019-12-10T13:26:34.99110940
BUILT $E135D5384EA42B4A771268CDA139D73D1C16A2C4~yutu,$066735D43FB476FA64578C5D829E176DF53E5A00~FreifunkIbbenbueren,$F248E8CEC8EAEA46E658DCA43500958FF21F4F46~manningisfree BUILD_FLAGS=IS_INTERNAL,NEED_CAPACITY PURPOSE=GENERAL TIME_CREATED=2019-12-10T13:26:35.99241141
BUILT $E135D5384EA42B4A771268CDA139D73D1C16A2C4~yutu,$F0C9513539800F762ECAE37F16370D7CBA5E52C2~Scrubs,$96DCD1D17C1F0FA4DED207C148E54C0A55553076~Honkler BUILD_FLAGS=IS_INTERNAL,NEED_CAPACITY PURPOSE=GENERAL TIME_CREATED=2019-12-10T13:26:36.99352842
BUILT $E135D5384EA42B4A771268CDA139D73D1C16A2C4~yutu,$5B1F0DAF378A1FAFCFD5FA9CDC66D1023DC0276E~fastlane,$0AD3B16ADF3EED3E5962FA944CD501352E790814~ritirong BUILD_FLAGS=IS_INTERNAL,NEED_CAPACITY PURPOSE=GENERAL TIME_CREATED=2019-12-10T13:26:37.99832043
BUILT $E135D5384EA42B4A771268CDA139D73D1C16A2C4~yutu,$303509AB910EF207B7438C27435C4A2FD579F1B1~Ukko,$13BC57BE4A6287D5C9AB970E94823BCAF7B5FFB2~OneTwoThree BUILD_FLAGS=IS_INTERNAL,NEED_CAPACITY PURPOSE=GENERAL TIME_CREATED=2019-12-10T13:26:39.00017044
BUILT $E135D5384EA42B4A771268CDA139D73D1C16A2C4~yutu,$16AE2704B794C6C0013C7E47AF11DCABA808A0B9~torreyporrey,$8ABC4D5F6182C7E79B087D1ED9E5CF652C83E52C~UDEtor BUILD_FLAGS=IS_INTERNAL,NEED_CAPACITY PURPOSE=GENERAL TIME_CREATED=2019-12-10T13:26:40.00053745
BUILT $E135D5384EA42B4A771268CDA139D73D1C16A2C4~yutu,$D8837125F99559B5385B4A09DD7007951DD17B92~Garados,$8A7F6489F82D564CB1EAD6A6762402696FCFF859~LapTor BUILD_FLAGS=IS_INTERNAL,NEED_CAPACITY PURPOSE=GENERAL TIME_CREATED=2019-12-10T13:26:41.00185046
BUILT $E135D5384EA42B4A771268CDA139D73D1C16A2C4~yutu,$F2DFE5FA1E4CF54F8E761A6D304B9B4EC69BDAE8~AlleKochenKaffee,$E00A55B52B7C933E715E30A892E1B27C7CC87EFF~rawpower BUILD_FLAGS=IS_INTERNAL,NEED_CAPACITY PURPOSE=GENERAL TIME_CREATED=2019-12-10T13:26:42.02174247
BUILT $E135D5384EA42B4A771268CDA139D73D1C16A2C4~yutu,$FFA72BD683BC2FCF988356E6BEC1E490F313FB07~Lule,$7C22E89AB620E77DF1B4A33787A5EC3F2AB4A8D1~ZKP1984 BUILD_FLAGS=IS_INTERNAL,NEED_CAPACITY PURPOSE=GENERAL TIME_CREATED=2019-12-10T13:26:43.026957"
```

perhaps we're doing some circuit isolation accidentally?

just opened up this ticket for me to track my research on it. going to see what circuit logic is like for [Tor Browser desktop](https://gitweb.torproject.org/tor-browser.git/) as well.

**Comment by mtigas** [comment]
copying over the doc about the control spec response line that we have here:  
https://torproject.gitlab.io/torspec/control-spec.html#circuit-status-changed

---

>  The syntax is:
> ```
>  "650" SP "CIRC" SP CircuitID SP CircStatus [SP Path]
>       [SP "BUILD_FLAGS=" BuildFlags] [SP "PURPOSE=" Purpose]
>       [SP "HS_STATE=" HSState] [SP "REND_QUERY=" HSAddress]
>       [SP "TIME_CREATED=" TimeCreated]
>       [SP "REASON=" Reason [SP "REMOTE_REASON=" Reason]]
>       [SP "SOCKS_USERNAME=" EscapedUsername]
>       [SP "SOCKS_PASSWORD=" EscapedPassword]
>       CRLF
> 
>   CircStatus =
>            "LAUNCHED" / ; circuit ID assigned to new circuit
>            "BUILT"    / ; all hops finished, can now accept streams
>            "GUARD_WAIT" / ; all hops finished, waiting to see if a
>                           ;  circuit with a better guard will be usable.
>            "EXTENDED" / ; one more hop has been completed
>            "FAILED"   / ; circuit closed (was not built)
>            "CLOSED"     ; circuit closed (was built)
> 
>   Path = LongName *("," LongName)
>     ; In Tor versions 0.1.2.2-alpha through 0.2.2.1-alpha with feature
>     ; VERBOSE_NAMES turned off and before version 0.1.2.2-alpha, Path
>     ; is as follows:
>     ; Path = ServerID *("," ServerID)
> 
>   BuildFlags = BuildFlag *("," BuildFlag)
>   BuildFlag = "ONEHOP_TUNNEL" / "IS_INTERNAL" /
>               "NEED_CAPACITY" / "NEED_UPTIME"
> 
>   Purpose = "GENERAL" / "HS_CLIENT_INTRO" / "HS_CLIENT_REND" /
>             "HS_SERVICE_INTRO" / "HS_SERVICE_REND" / "TESTING" /
>             "CONTROLLER" / "MEASURE_TIMEOUT"
> 
>   HSState = "HSCI_CONNECTING" / "HSCI_INTRO_SENT" / "HSCI_DONE" /
>             "HSCR_CONNECTING" / "HSCR_ESTABLISHED_IDLE" /
>             "HSCR_ESTABLISHED_WAITING" / "HSCR_JOINED" /
>             "HSSI_CONNECTING" / "HSSI_ESTABLISHED" /
>             "HSSR_CONNECTING" / "HSSR_JOINED"
> 
>   EscapedUsername = QuotedString
>   EscapedPassword = QuotedString
> 
>   HSAddress = 16*Base32Character / 56*Base32Character
>   Base32Character = ALPHA / "2" / "3" / "4" / "5" / "6" / "7"
> 
>   TimeCreated = ISOTime2Frac
>   Seconds = 1*DIGIT
>   Microseconds = 1*DIGIT
> 
>   Reason = "NONE" / "TORPROTOCOL" / "INTERNAL" / "REQUESTED" /
>            "HIBERNATING" / "RESOURCELIMIT" / "CONNECTFAILED" /
>            "OR_IDENTITY" / "OR_CONN_CLOSED" / "TIMEOUT" /
>            "FINISHED" / "DESTROYED" / "NOPATH" / "NOSUCHSERVICE" /
>            "MEASUREMENT_EXPIRED"
> ```
>
> The path is provided only when the circuit has been extended at least one hop.
> 
> The “BUILD_FLAGS” field is provided only in versions 0.2.3.11-alpha and later. Clients MUST accept build flags not listed above. Build flags are defined as follows:
> 
> ```
>   ONEHOP_TUNNEL   (one-hop circuit, used for tunneled directory conns)
>   IS_INTERNAL     (internal circuit, not to be used for exiting streams)
>   NEED_CAPACITY   (this circuit must use only high-capacity nodes)
>   NEED_UPTIME     (this circuit must use only high-uptime nodes)
> ```
>
> The “PURPOSE” field is provided only in versions 0.2.1.6-alpha and later, and only if extended events are enabled (see 3.19). Clients MUST accept purposes not listed above. Purposes are defined as follows:
> 
> ```
>   GENERAL         (circuit for AP and/or directory request streams)
>   HS_CLIENT_INTRO (HS client-side introduction-point circuit)
>   HS_CLIENT_REND  (HS client-side rendezvous circuit; carries AP streams)
>   HS_SERVICE_INTRO (HS service-side introduction-point circuit)
>   HS_SERVICE_REND (HS service-side rendezvous circuit)
>   TESTING         (reachability-testing circuit; carries no traffic)
>   CONTROLLER      (circuit built by a controller)
>   MEASURE_TIMEOUT (circuit being kept around to see how long it takes)
> ```
>
> The “HS_STATE” field is provided only for hidden-service circuits, and only in versions 0.2.3.11-alpha and later. Clients MUST accept hidden-service circuit states not listed above. Hidden-service circuit states are defined as follows:
> 
> ```
>   HSCI_*      (client-side introduction-point circuit states)
>     HSCI_CONNECTING          (connecting to intro point)
>     HSCI_INTRO_SENT          (sent INTRODUCE1; waiting for reply from IP)
>     HSCI_DONE                (received reply from IP relay; closing)
> 
>   HSCR_*      (client-side rendezvous-point circuit states)
>     HSCR_CONNECTING          (connecting to or waiting for reply from RP)
>     HSCR_ESTABLISHED_IDLE    (established RP; waiting for introduction)
>     HSCR_ESTABLISHED_WAITING (introduction sent to HS; waiting for rend)
>     HSCR_JOINED              (connected to HS)
> 
>   HSSI_*      (service-side introduction-point circuit states)
>     HSSI_CONNECTING          (connecting to intro point)
>     HSSI_ESTABLISHED         (established intro point)
> 
>   HSSR_*      (service-side rendezvous-point circuit states)
>     HSSR_CONNECTING          (connecting to client's rend point)
>     HSSR_JOINED              (connected to client's RP circuit)
> ```
> 
> The “SOCKS_USERNAME” and “SOCKS_PASSWORD” fields indicate the credentials that were used by a SOCKS client to connect to Tor’s SOCKS port and initiate this circuit. (Streams for SOCKS clients connected with different usernames and/or passwords are isolated on separate circuits if the IsolateSOCKSAuth flag is active; see Proposal 171.) [Added in Tor 0.4.3.1-alpha.]
> 
> The “REND_QUERY” field is provided only for hidden-service-related circuits, and only in versions 0.2.3.11-alpha and later. Clients MUST accept hidden service addresses in formats other than that specified above. [Added in Tor 0.4.3.1-alpha.]
> 
> The “TIME_CREATED” field is provided only in versions 0.2.3.11-alpha and later. TIME_CREATED is the time at which the circuit was created or cannibalized. [Added in Tor 0.4.3.1-alpha.]
> 
> The “REASON” field is provided only for FAILED and CLOSED events, and only if extended events are enabled (see 3.19). Clients MUST accept reasons not listed above. [Added in Tor 0.4.3.1-alpha.] Reasons are as given in tor-spec.txt, except for:
> 
> ```
>   NOPATH              (Not enough nodes to make circuit)
>   MEASUREMENT_EXPIRED (As "TIMEOUT", except that we had left the circuit
>                        open for measurement purposes to see how long it
>                        would take to finish.)
> ```
> 
> The “REMOTE_REASON” field is provided only when we receive a DESTROY or TRUNCATE cell, and only if extended events are enabled. It contains the actual reason given by the remote OR for closing the circuit. Clients MUST accept reasons not listed above. Reasons are as listed in tor-spec.txt. [Added in Tor 0.4.3.1-alpha.]
>

**Comment by mtigas** [comment]
Looks like TBB does do a thing where it sets a SOCKS5 username/password for each combination of `tab + first-party domain` (as in, the domain of the page that is being viewed; vs domains of assets -- `<img>`, etc -- that are on that page)

https://github.com/mtigas/torbutton/blob/ed3fafebab28d005e22208aa921e5f5c7c54153f/chrome/content/tor-circuit-display.js#L43-L70

https://github.com/mtigas/torbutton/blob/ed3fafebab28d005e22208aa921e5f5c7c54153f/chrome/content/tor-circuit-display.js#L263-L279

**Comment by n8fr8** [comment]
Yes if you randomly generate user/pass for each connection, it will make a new circuit

**Comment by n8fr8** [comment]
(and if you keep them the same per tab, domain, then it should use the same circuit as before)

**Comment by tladesignz** [comment]
> Yes if you randomly generate user/pass for each connection, it will make a new circuit

Pretty sure, we don‘t do that.

**Comment by mtigas** [comment]
tl;dr --- I've experimented & given this a bit of thought. I wonder if we should be using `NoIsolateDestPort NoIsolateDestAddr` and using _a single circuit_ for the app (until we properly figure out #209). I think it might be better browser behavior (but am not 100% sure), and it would solve the "which circuit are we using" problem.

---

Did a quick test here and confirmed I see similar behavior as @tladesignz does from the the `circuit-status` output.  I changed the `socksport` line in OnionManager.swift to this:
```
            "--socksport", "39050 IsolateSOCKSAuth NoIsolateDestPort NoIsolateDestAddr",
```

And then this bit in JAHPAuthenticatingHTTPProtocol.m (adding the `User` and `Password` lines):

```
	if (socksProxyPort > 0) {
		proxyDict[@"SOCKSEnable"] = @YES;
		proxyDict[(NSString *)kCFStreamPropertySOCKSProxyHost] = proxyHost;
		proxyDict[(NSString *)kCFStreamPropertySOCKSProxyPort] = [NSNumber numberWithInteger: socksProxyPort];
		proxyDict[(NSString *)kCFStreamPropertySOCKSUser] = @"onionbrowser";
		proxyDict[(NSString *)kCFStreamPropertySOCKSPassword] = @"asdfid";
	}
```

My `circuit-status` output now looks like this:

```
1 BUILT $43FF3346A3F8E4ED27F631DC956C4CD8D6173D7A~fujiwara,$127BA287DCA8F7ECF1D4021D81AA5500E1B6517F~ujhlfhgd,$63799328222647BE7ED9171FC9FDDED0B77AAFB5~GetYourAssToMars BUILD_FLAGS=NEED_CAPACITY PURPOSE=GENERAL TIME_CREATED=2019-12-10T23:13:21.882368 SOCKS_USERNAME=\"onionbrowser\" SOCKS_PASSWORD=\"asdfid\"
2 BUILT $43FF3346A3F8E4ED27F631DC956C4CD8D6173D7A~fujiwara,$088A6237EED87EDEB0E1E3AD0A4A419F963F0787~Unnamed,$B84F248233FEA90CAD439F292556A3139F6E1B82~tollana BUILD_FLAGS=NEED_CAPACITY PURPOSE=GENERAL TIME_CREATED=2019-12-10T23:13:22.886659
3 BUILT $43FF3346A3F8E4ED27F631DC956C4CD8D6173D7A~fujiwara,$4CA64E677ECFF23B138CFE70A6A02E94BDF3A805~grumpytechie1,$AD7C289DBCEE8522462AD9E3BDF25E2576FC8F30~ToolHaven9cf,$1F509589F7F70B69A38719A201451CF4B70F89C6~CulNoir BUILD_FLAGS=IS_INTERNAL,NEED_CAPACITY PURPOSE=HS_CLIENT_HSDIR HS_STATE=HSCI_CONNECTING TIME_CREATED=2019-12-10T23:13:23.505373
4 BUILT $43FF3346A3F8E4ED27F631DC956C4CD8D6173D7A~fujiwara,$1D3D0DC349A29D53FF890F9DACBDD1E7EA37F5B0~EarthElephant1,$4EBE3B18F7D38D0A2FDD67B47A55CB9069368723~ForJusticeDammit BUILD_FLAGS=IS_INTERNAL,NEED_CAPACITY PURPOSE=HS_CLIENT_REND HS_STATE=HSCR_JOINED REND_QUERY=3heens4xbedlj57xwcggjsdglot7e36p4rogy642xokemfo2duh6bbyd TIME_CREATED=2019-12-10T23:13:23.887722 SOCKS_USERNAME=\"onionbrowser\" SOCKS_PASSWORD=\"asdfid\"
5 BUILT $43FF3346A3F8E4ED27F631DC956C4CD8D6173D7A~fujiwara,$B91A1EB30E66D52026EFCD1E96BFC0E966C83E6D~AssangeFrB8too,$24A17A4C1FA8C2108ACDE3BE681B0731384BF3A5~dragonhoard1 BUILD_FLAGS=IS_INTERNAL,NEED_CAPACITY PURPOSE=GENERAL TIME_CREATED=2019-12-10T23:13:24.889375
7 BUILT $43FF3346A3F8E4ED27F631DC956C4CD8D6173D7A~fujiwara,$188BE26E437D0B0BDD90E0717558CA686D288BAA~cloudtunnel,$8C5B316ED73018484765C3E0944E4508DCAE0944~Waldmeister2 BUILD_FLAGS=IS_INTERNAL,NEED_CAPACITY PURPOSE=GENERAL TIME_CREATED=2019-12-10T23:13:25.894132
8 BUILT $43FF3346A3F8E4ED27F631DC956C4CD8D6173D7A~fujiwara,$E45342180C2C7A84981E163751F8981A17585C17~Unnamed,$52EEBEAA873C1520915BBE661520F2DC197ED84A~FreeRussia BUILD_FLAGS=IS_INTERNAL,NEED_CAPACITY PURPOSE=GENERAL TIME_CREATED=2019-12-10T23:13:26.897041
9 BUILT $43FF3346A3F8E4ED27F631DC956C4CD8D6173D7A~fujiwara,$F6740DEABFD5F62612FA025A5079EA72846B1F67~poiuty,$E6C04728E9339DF22F331B7017D4A7AA3F252F0B~malekalcom BUILD_FLAGS=IS_INTERNAL,NEED_CAPACITY PURPOSE=GENERAL TIME_CREATED=2019-12-10T23:13:27.898952
10 BUILT $43FF3346A3F8E4ED27F631DC956C4CD8D6173D7A~fujiwara,$841DE1E065BCA8C1EBDC4A7140646F47A6257BD8~digitalResistanceRu,$92173C9F108197460307969D353333DACE9B9F09~bgsv BUILD_FLAGS=IS_INTERNAL,NEED_CAPACITY PURPOSE=GENERAL TIME_CREATED=2019-12-10T23:13:28.899613
11 BUILT $43FF3346A3F8E4ED27F631DC956C4CD8D6173D7A~fujiwara,$DB8283A4E152010A9CAFDA63CC0F56ACA1556769~topinambour2,$57C6DF5B93E54EB9C8DB90029D9E9A1111BD34D2~angeltest12 BUILD_FLAGS=IS_INTERNAL,NEED_CAPACITY PURPOSE=GENERAL TIME_CREATED=2019-12-10T23:13:29.900674
12 BUILT $43FF3346A3F8E4ED27F631DC956C4CD8D6173D7A~fujiwara,$D8A2D0080B0C73E1DC0EC9F33B497F64E5FB3F52~porthill,$9E9BEC429AC8FB99795ACDCEE09BC56049348ECB~nostradamus BUILD_FLAGS=IS_INTERNAL,NEED_CAPACITY PURPOSE=GENERAL TIME_CREATED=2019-12-10T23:13:30.901640
13 BUILT $43FF3346A3F8E4ED27F631DC956C4CD8D6173D7A~fujiwara,$34DE7F9028A53D5238D779B00DAF7438FDE1D7A0~ishouldeditheconfi,$C2C2B839F4CFFA5F7B7A40CB0BE84F7D948CB510~behrmann BUILD_FLAGS=IS_INTERNAL,NEED_CAPACITY PURPOSE=GENERAL TIME_CREATED=2019-12-10T23:13:31.902913
14 BUILT $43FF3346A3F8E4ED27F631DC956C4CD8D6173D7A~fujiwara,$1CD48F4ED0F1821FFBF1940802A13EEFD4C27502~Piratenpartei00,$B2EFB9FF778A4CA2F5895582B52CD3757D455355~Unnamed BUILD_FLAGS=IS_INTERNAL,NEED_CAPACITY PURPOSE=GENERAL TIME_CREATED=2019-12-10T23:13:32.905982
19 BUILT $43FF3346A3F8E4ED27F631DC956C4CD8D6173D7A~fujiwara,$2007B35F41D248DABFA11A2A4AA0B10517801853~falcon,$311A4533F7A2415F42346A6C8FA77E6FD279594C~DigiGesTor3e2 BUILD_FLAGS=NEED_CAPACITY PURPOSE=GENERAL TIME_CREATED=2019-12-10T23:13:45.934839
16 BUILT $43FF3346A3F8E4ED27F631DC956C4CD8D6173D7A~fujiwara,$EFE449E0B6B7A657376B5FE62AD172B3EC5800A8~uppsalarelay,$B32B93C90DCAF0012328A8D55D621D26B186E11A~OnionOnion123 BUILD_FLAGS=IS_INTERNAL,NEED_CAPACITY PURPOSE=GENERAL TIME_CREATED=2019-12-10T23:13:34.908258
17 BUILT $43FF3346A3F8E4ED27F631DC956C4CD8D6173D7A~fujiwara,$5CFC486780CBD4446B764E51608D6574003B350D~helena,$380B9F1533AAD587F4F6E3E7AAD661F8DBBA246E~Bahar BUILD_FLAGS=IS_INTERNAL,NEED_CAPACITY PURPOSE=GENERAL TIME_CREATED=2019-12-10T23:13:35.911844
18 BUILT $43FF3346A3F8E4ED27F631DC956C4CD8D6173D7A~fujiwara,$1949BAA1F64F01331DFBD32403303688FEBC05A0~Unnamed,$3D9CE80399B68C01377FEC42395E399B9ED16E18~megatot BUILD_FLAGS=IS_INTERNAL,NEED_CAPACITY PURPOSE=GENERAL TIME_CREATED=2019-12-10T23:13:36.912919
```

So certainly all the Safari traffic is using circuit 1, since the UIWebView is configured to set that SOCK5 user/pass. (Except for when visiting the homepage `.onion`. I need to read the spec some more, but I think it may be using circuit 4 for that one, since that's the `HS_CLIENT_REND` connection to the rendezvous point for the onion service, in the six-hop link that happens with onion serices.)

Basically the config makes it so that _we_ have to handle which things share circuits (set `NoIsolateDestPort NoIsolateDestAddr`), rather than Tor's default behavior of every remote host and port getting a different circuit.

(I _think_ implementing proper isolation bits would actually fix this problem too. I will discuss that part of actually giving a dynamic SOCKS5 user/pass -- instead of `"asdfid"` -- in #209.)

I think we should have always been setting `NoIsolateDestPort NoIsolateDestAddr` on the SocksPort line, tbh. The current behavior is bad b/c we get many circuits within one website (since Tor would open a new one for every host), which does make the traffic look suspicious and also bloat us with excessive circuits. That behavior is good for a neutral proxy that is unaware of the implementing application, but probably not for a web browser.

(It's one of those behaviors that I recall being discussed when folks use the tor proxy and use Chrome or something else over it, rather than use Tor Browser. It's not recommended for those reasons. I also think it may be breaking reCAPTCHAs due to how they are hosted cross-origin by Google. Like we should probably be using But that's a whole other story.))

---

@n8fr8 Do you have any thoughts on whether it'd be a good or bad idea to use `NoIsolateDestPort NoIsolateDestAddr`? (So it'd be a single circuit app-wide -- except for onion services which are handled specially by tor --  until we actually solve #209.) I will also probably ask in Tor IRC later tonight.

**Comment by mtigas** [comment]
If we decide we don't do `NoIsolateDestPort NoIsolateDestAddr`, then I don't think we should display circuit info until #209 is done (since it wouldn't be accurate).

**Comment by tladesignz** [comment]
Ok, so I finally developed a full parser for Tor circuits and its own class to store the results:
https://github.com/iCepa/Tor.framework/commit/ba886329ebf085c7ec1dd8bf0fceb89c26d2fc35

Now we need the rules to decide, which circuit was actually used.

- If we leverage the SOCKS user, then we filter by that. Do we actually need that?
- If it's an `.onion` address, we filter by that in `REND_QUERY`.
- Probably also filter by `PURPOSE`?
- I noticed, a lot of the circuits have `BUILD_FLAG=IS_INTERNAL`. Can we filter these out? Although, the one with a `REND_QUERY` for the `.onion` address also has that and another circuit didn't when I tested. Hmmm. Guess we can't do this without tagging it with a SOCKS user?

**Comment by tladesignz** [comment]
Ok, so thanks to @mtigas insights, I think I have this covered, basically:

- When an onion site, I search for the circuit doing the rendezvous with that site. That seems to always be a resultset of one, so quite correct.

- When a normal site, I filter by purpose (should be `GENERAL`), by build flags (should not contain `IS_INTERNAL` and `ONE_HOP_TUNNEL`) and I tag our requests with a SOCKS username/password which I also filter on.
That gives me 2 - 4 circuits, which I reverse order by `TIME_CREATED` and I take the newest.
That seems to do it mostly. However, it looks like there are situations, when a new circuit is created right after a request. In that case, I show the wrong circuit and I have no idea how I could further identify the correct circuit. I guess, that will also happen, if we would do tab isolation. Maybe that's the reason why Tor Browser Desktop did display the wrong circuit once, when I tried.

This experiment was conducted without `NoIsolateDestPort` and `NoIsolateDestAddr`.
(BTW: `IsolateSOCKSAuth` is default as per spec.)

Anyway, it seems, I'm switching circuits quite often with this setup. Not between requests (I get the same IP on multiple different IP-address-reflection sites), but just randomly every 5 minutes or so.

IP reflection sites that work over Tor without Captchas:

http://showip.net/
https://www.wieistmeineip.de/
https://www.showmyip.gr/

**Comment by tladesignz** [comment]
Aaahaaa. From https://stackoverflow.com/a/32347493 :

> If you run a TOR relay the signal NEWNYM has no effect. it works only for TOR clients for browsing the internet. It means new identity, new circuit.
> 
> If Tor chose to ignore a signal (such as NEWNYM), this event will not be sent. Note that some options (like ReloadTorrcOnSIGHUP) may affect the semantics of the signals here.
> 
> Note that the HALT (SIGTERM) and SHUTDOWN (SIGINT) signals do not currently generate any event.

Currently NEWNYM doesn't have any effect on me. Could it be that our Tor thinks it's running in relay mode when it should run in client mode? How to tell Tor that?

**Comment by tladesignz** [comment]
Urgh. Seems, circuits only change after an additional page reload.
Looking at what Tor Browser does, I guess they do a CLOSECIRCUIT of all circuits they identified as being used by that tab. It could have been so nice...

**Comment by tladesignz** [comment]
Ok, it seems, it actually *is* normal. The bulk of these circuits are actually *internal*. Since I basically nailed #223, I'll close this, as it doesn't make sense to me to have this laying around.
